modules/kameleonvolume/rendering/renderablekameleonvolume.cpp

```diff
@@ -0,0 +1,375 @@
+    );
+    if (FileSys.fileExists(cachePath)) {
+        loadRaw(cachePath);
+    } else {
+        loadFromPath(_sourcePath);
+        storeRaw(cachePath);
+    }
+}
+
+std::string RenderableKameleonVolume::cacheSuffix() {
+    glm::vec3 dims = _dimensions.value();
+    return "." + _variable.value() +
+        "." + std::to_string(dims[0]) +
+        "x" + std::to_string(dims[1]) +
+        "x" + std::to_string(dims[2]);
+}
+
+void RenderableKameleonVolume::loadFromPath(const std::string& path) {
+    ghoul::filesystem::File file(path);
+    std::string extension = file.fileExtension();
+    std::transform(extension.begin(), extension.end(), extension.begin(), ::tolower);
+    if (extension == "cdf") {
+        loadCdf(path);
+    } else {
+        loadRaw(path);
+    }
+}
+
+void RenderableKameleonVolume::loadRaw(const std::string& path) {
+    RawVolumeReader<float> reader(path, _dimensions);
+    _rawVolume = reader.read();
+    updateTextureFromVolume();
+}
+
+void RenderableKameleonVolume::loadCdf(const std::string& path) {
+    KameleonVolumeReader reader(path);
+
+    if (_autoValueBounds) {
+        _lowerValueBound = reader.minValue(_variable);
+        _upperValueBound = reader.maxValue(_variable);
+    }
+
+    std::vector<std::string> variables = reader.gridVariableNames();
+   
+    if (variables.size() == 3 && _autoDomainBounds) {
+        _lowerDomainBound = glm::vec3(
+            reader.minValue(variables[0]),
+            reader.minValue(variables[1]),
+            reader.minValue(variables[2]));
+
+        _upperDomainBound = glm::vec3(
+            reader.maxValue(variables[0]),
+            reader.maxValue(variables[1]),
+            reader.maxValue(variables[2]));
+    }
+
+    if (variables.size() == 3 && _autoGridType) {
+        if (variables[0] == "r" && variables[0] == "theta" && variables[0] == "phi") {
+            _gridType.setValue(static_cast<int>(VolumeGridType::Spherical));
+        }
+        else {
+            _gridType.setValue(static_cast<int>(VolumeGridType::Cartesian));
+        }       
+    }
+
+    ghoul::Dictionary dict = reader.readMetaData();
+    _rawVolume = reader.readFloatVolume(_dimensions, _variable, _lowerDomainBound, _upperDomainBound);
+    updateTextureFromVolume();
+}
+
+void RenderableKameleonVolume::updateTextureFromVolume() {
+    _normalizedVolume = std::make_unique<RawVolume<GLfloat>>(_dimensions);
+    float* in = _rawVolume->data();
+    GLfloat* out = _normalizedVolume->data();
+    float min = _lowerValueBound;
+    float diff = _upperValueBound - _lowerValueBound;
+
+    for (size_t i = 0; i < _normalizedVolume->nCells(); i++) {
+        out[i] = glm::clamp((in[i] - min) / diff, 0.0f, 1.0f);
+    }
+
+    _volumeTexture = std::make_shared<ghoul::opengl::Texture>(
+        _dimensions,
+        ghoul::opengl::Texture::Format::Red,
+        GL_RED,
+        GL_FLOAT,
+        ghoul::opengl::Texture::FilterMode::Linear,
+        ghoul::opengl::Texture::WrappingMode::Repeat);
```

Comment:
for consistency:

```
_volumeTexture = std::make_shared<ghoul::opengl::Texture>(
    _dimensions,
    ghoul::opengl::Texture::Format::Red,
    GL_RED,
    GL_FLOAT,
    ghoul::opengl::Texture::FilterMode::Linear,
    ghoul::opengl::Texture::WrappingMode::Repeat
);
```

reason:  Enforce an extra line to visually separate the complex function call from the following lines